# 112-1 ML_HW2_Group homework
* Group: 3
* Member: 邱寶樟(26)、侯亮羽(14)、李奕呈(18)

    We will use 5 categorical features and 11 numerical features to predict y. 
    You need to follow most of the questions as in the individual homework, but emphasize the following specific questions.

## (a)
* Please use the function get_dummies() to transform the categorical features into 0 and 1 numbers.<br> 
* Inside the get_dummies() function, you could provide all of the features which is easier than the approach in the lecture note. That is, the numerical features.

In [1]:
import pandas as pd 
df = pd.read_csv("throughput.csv")
df.head()

Y    X1    X2    X3    X4    X5        X6   X7   X8   X9  X10  \
0  1546.836248  X1-1  X2-1  X3-1  X4-1  X5-1  0.025286   50   48  105  3.0   
1  1261.722593  X1-1  X2-1  X3-1  X4-1  X5-1  0.034876   67   34  105  3.0   
2   860.853275  X1-1  X2-1  X3-1  X4-1  X5-1  0.079237  176  187  200  8.0   
3   887.356349  X1-1  X2-1  X3-1  X4-1  X5-1  0.113288  176  187  200  8.0   
4   895.059756  X1-1  X2-1  X3-1  X4-1  X5-1  0.113778  176  187  200  8.0   

   X11   X12  X13   X14  X15  X16  
0  2.0  0.45    8  1600    8  0.8  
1  2.0  0.45    8  1600   10  0.8  
2  6.0  0.75    8   216   10  0.8  
3  6.0  0.75    8   216   10  0.8  
4  6.0  0.75    8   216   10  0.8

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13186 entries, 0 to 13185
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Y       13186 non-null  float64
 1   X1      13186 non-null  object 
 2   X2      13186 non-null  object 
 3   X3      13186 non-null  object 
 4   X4      13185 non-null  object 
 5   X5      13186 non-null  object 
 6   X6      13186 non-null  float64
 7   X7      13186 non-null  int64  
 8   X8      13186 non-null  int64  
 9   X9      13186 non-null  int64  
 10  X10     13186 non-null  float64
 11  X11     13186 non-null  float64
 12  X12     13186 non-null  float64
 13  X13     13186 non-null  int64  
 14  X14     13186 non-null  int64  
 15  X15     13186 non-null  int64  
 16  X16     13186 non-null  float64
dtypes: float64(6), int64(6), object(5)
memory usage: 1.7+ MB


In [3]:
throughput = df.dropna()
throughput.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13185 entries, 0 to 13185
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Y       13185 non-null  float64
 1   X1      13185 non-null  object 
 2   X2      13185 non-null  object 
 3   X3      13185 non-null  object 
 4   X4      13185 non-null  object 
 5   X5      13185 non-null  object 
 6   X6      13185 non-null  float64
 7   X7      13185 non-null  int64  
 8   X8      13185 non-null  int64  
 9   X9      13185 non-null  int64  
 10  X10     13185 non-null  float64
 11  X11     13185 non-null  float64
 12  X12     13185 non-null  float64
 13  X13     13185 non-null  int64  
 14  X14     13185 non-null  int64  
 15  X15     13185 non-null  int64  
 16  X16     13185 non-null  float64
dtypes: float64(6), int64(6), object(5)
memory usage: 1.8+ MB


In [4]:
throughput.iloc[:, 1:6]

X1    X2    X3    X4     X5
0      X1-1  X2-1  X3-1  X4-1   X5-1
1      X1-1  X2-1  X3-1  X4-1   X5-1
2      X1-1  X2-1  X3-1  X4-1   X5-1
3      X1-1  X2-1  X3-1  X4-1   X5-1
4      X1-1  X2-1  X3-1  X4-1   X5-1
...     ...   ...   ...   ...    ...
13181  X1-2  X2-2  X3-2  X4-7  X5-17
13182  X1-2  X2-2  X3-2  X4-7  X5-17
13183  X1-2  X2-2  X3-2  X4-7  X5-17
13184  X1-2  X2-2  X3-1  X4-7  X5-21
13185  X1-2  X2-2  X3-1  X4-7  X5-21

[13185 rows x 5 columns]

In [5]:
pd.get_dummies(throughput.iloc[:, 1:6]).head()

X1_X1-1  X1_X1-2  X2_X2-1  X2_X2-2  X2_X2-3  X3_X3-1  X3_X3-2  X3_X3-3  \
0        1        0        1        0        0        1        0        0   
1        1        0        1        0        0        1        0        0   
2        1        0        1        0        0        1        0        0   
3        1        0        1        0        0        1        0        0   
4        1        0        1        0        0        1        0        0   

   X3_X3-4  X4_X4-1  ...  X5_X5-20  X5_X5-21  X5_X5-22  X5_X5-3  X5_X5-4  \
0        0        1  ...         0         0         0        0        0   
1        0        1  ...         0         0         0        0        0   
2        0        1  ...         0         0         0        0        0   
3        0        1  ...         0         0         0        0        0   
4        0        1  ...         0         0         0        0        0   

   X5_X5-5  X5_X5-6  X5_X5-7  X5_X5-8  X5_X5-9  
0        0        0        0        0        0  
1        0        0        0        0        0  
2        0        0        0        0        0  
3        0        0        0        0        0  
4        0        0        0        0        0  

[5 rows x 38 columns]

In [6]:
pd.get_dummies(throughput.iloc[:, 1:6]).astype(int).head()

X1_X1-1  X1_X1-2  X2_X2-1  X2_X2-2  X2_X2-3  X3_X3-1  X3_X3-2  X3_X3-3  \
0        1        0        1        0        0        1        0        0   
1        1        0        1        0        0        1        0        0   
2        1        0        1        0        0        1        0        0   
3        1        0        1        0        0        1        0        0   
4        1        0        1        0        0        1        0        0   

   X3_X3-4  X4_X4-1  ...  X5_X5-20  X5_X5-21  X5_X5-22  X5_X5-3  X5_X5-4  \
0        0        1  ...         0         0         0        0        0   
1        0        1  ...         0         0         0        0        0   
2        0        1  ...         0         0         0        0        0   
3        0        1  ...         0         0         0        0        0   
4        0        1  ...         0         0         0        0        0   

   X5_X5-5  X5_X5-6  X5_X5-7  X5_X5-8  X5_X5-9  
0        0        0        0        0        0  
1        0        0        0        0        0  
2        0        0        0        0        0  
3        0        0        0        0        0  
4        0        0        0        0        0  

[5 rows x 38 columns]

## (b)
* What are the most important correlation coefficients in terms of absolute magnitude with respect to the column Y? 
* Does it make a difference as compared with the numerical features only?

In [7]:
categorical = pd.get_dummies(throughput.iloc[:, 1:])
categorical

X6   X7   X8   X9   X10   X11   X12  X13   X14  X15  ...  \
0      0.025286   50   48  105   3.0   2.0  0.45    8  1600    8  ...   
1      0.034876   67   34  105   3.0   2.0  0.45    8  1600   10  ...   
2      0.079237  176  187  200   8.0   6.0  0.75    8   216   10  ...   
3      0.113288  176  187  200   8.0   6.0  0.75    8   216   10  ...   
4      0.113778  176  187  200   8.0   6.0  0.75    8   216   10  ...   
...         ...  ...  ...  ...   ...   ...   ...  ...   ...  ...  ...   
13181  0.149699  479  247   60  12.0  18.0  1.20  132    60   88  ...   
13182  0.300309  217  216   80   8.0  10.5  1.00  162   168  126  ...   
13183  0.300309  217  216   80   8.0  10.5  1.00  162   168   26  ...   
13184  0.154297  370  327  150  10.0  14.5  0.80  200    70  182  ...   
13185  0.163640  370  327  150  10.0  14.5  0.80  200    70  182  ...   

       X5_X5-20  X5_X5-21  X5_X5-22  X5_X5-3  X5_X5-4  X5_X5-5  X5_X5-6  \
0             0         0         0        0        0        0        0   
1             0         0         0        0        0        0        0   
2             0         0         0        0        0        0        0   
3             0         0         0        0        0        0        0   
4             0         0         0        0        0        0        0   
...         ...       ...       ...      ...      ...      ...      ...   
13181         0         0         0        0        0        0        0   
13182         0         0         0        0        0        0        0   
13183         0         0         0        0        0        0        0   
13184         0         1         0        0        0        0        0   
13185         0         1         0        0        0        0        0   

       X5_X5-7  X5_X5-8  X5_X5-9  
0            0        0        0  
1            0        0        0  
2            0        0        0  
3            0        0        0  
4            0        0        0  
...        ...      ...      ...  
13181        0        0        0  
13182        0        0        0  
13183        0        0        0  
13184        0        0        0  
13185        0        0        0  

[13185 rows x 49 columns]

In [8]:
from sklearn.preprocessing import MinMaxScaler
ss = MinMaxScaler()
X_scaled = ss.fit_transform(categorical)
X_scaled.shape

(13185, 49)

In [9]:
pd.DataFrame(X_scaled, columns = categorical.columns[0:])

X6        X7        X8        X9       X10       X11       X12  \
0      0.001773  0.057196  0.055133  0.078571  0.026316  0.000000  0.030864   
1      0.002447  0.088561  0.028517  0.078571  0.026316  0.000000  0.030864   
2      0.005567  0.289668  0.319392  0.214286  0.157895  0.105263  0.216049   
3      0.007962  0.289668  0.319392  0.214286  0.157895  0.105263  0.216049   
4      0.007997  0.289668  0.319392  0.214286  0.157895  0.105263  0.216049   
...         ...       ...       ...       ...       ...       ...       ...   
13180  0.010523  0.848708  0.433460  0.014286  0.263158  0.421053  0.493827   
13181  0.021117  0.365314  0.374525  0.042857  0.157895  0.223684  0.370370   
13182  0.021117  0.365314  0.374525  0.042857  0.157895  0.223684  0.370370   
13183  0.010847  0.647601  0.585551  0.142857  0.210526  0.328947  0.246914   
13184  0.011504  0.647601  0.585551  0.142857  0.210526  0.328947  0.246914   

            X13       X14       X15  ...  X5_X5-20  X5_X5-21  X5_X5-22  \
0      0.000000  0.666110  0.010526  ...       0.0       0.0       0.0   
1      0.000000  0.666110  0.015789  ...       0.0       0.0       0.0   
2      0.000000  0.088481  0.015789  ...       0.0       0.0       0.0   
3      0.000000  0.088481  0.015789  ...       0.0       0.0       0.0   
4      0.000000  0.088481  0.015789  ...       0.0       0.0       0.0   
...         ...       ...       ...  ...       ...       ...       ...   
13180  0.323760  0.023372  0.221053  ...       0.0       0.0       0.0   
13181  0.402089  0.068447  0.321053  ...       0.0       0.0       0.0   
13182  0.402089  0.068447  0.057895  ...       0.0       0.0       0.0   
13183  0.501305  0.027546  0.468421  ...       0.0       1.0       0.0   
13184  0.501305  0.027546  0.468421  ...       0.0       1.0       0.0   

       X5_X5-3  X5_X5-4  X5_X5-5  X5_X5-6  X5_X5-7  X5_X5-8  X5_X5-9  
0          0.0      0.0      0.0      0.0      0.0      0.0      0.0  
1          0.0      0.0      0.0      0.0      0.0      0.0      0.0  
2          0.0      0.0      0.0      0.0      0.0      0.0      0.0  
3          0.0      0.0      0.0      0.0      0.0      0.0      0.0  
4          0.0      0.0      0.0      0.0      0.0      0.0      0.0  
...        ...      ...      ...      ...      ...      ...      ...  
13180      0.0      0.0      0.0      0.0      0.0      0.0      0.0  
13181      0.0      0.0      0.0      0.0      0.0      0.0      0.0  
13182      0.0      0.0      0.0      0.0      0.0      0.0      0.0  
13183      0.0      0.0      0.0      0.0      0.0      0.0      0.0  
13184      0.0      0.0      0.0      0.0      0.0      0.0      0.0  

[13185 rows x 49 columns]

In [10]:
pd.DataFrame(throughput['Y'], columns = [throughput.columns[0]])

Y
0      1546.836248
1      1261.722593
2       860.853275
3       887.356349
4       895.059756
...            ...
13181   129.837472
13182     7.835104
13183     7.835104
13184   128.375469
13185    74.586662

[13185 rows x 1 columns]

In [11]:
throughput['Y'].shape, X_scaled.shape

((13185,), (13185, 49))

In [12]:
data = pd.concat([pd.DataFrame(X_scaled, columns = categorical.columns[ 0:]), 
                  pd.DataFrame(throughput['Y'].values, columns = [throughput.columns[0]])], axis = 1) 
data.head() # See the result and make sure that the above command works!

X6        X7        X8        X9       X10       X11       X12  X13  \
0  0.001773  0.057196  0.055133  0.078571  0.026316  0.000000  0.030864  0.0   
1  0.002447  0.088561  0.028517  0.078571  0.026316  0.000000  0.030864  0.0   
2  0.005567  0.289668  0.319392  0.214286  0.157895  0.105263  0.216049  0.0   
3  0.007962  0.289668  0.319392  0.214286  0.157895  0.105263  0.216049  0.0   
4  0.007997  0.289668  0.319392  0.214286  0.157895  0.105263  0.216049  0.0   

        X14       X15  ...  X5_X5-21  X5_X5-22  X5_X5-3  X5_X5-4  X5_X5-5  \
0  0.666110  0.010526  ...       0.0       0.0      0.0      0.0      0.0   
1  0.666110  0.015789  ...       0.0       0.0      0.0      0.0      0.0   
2  0.088481  0.015789  ...       0.0       0.0      0.0      0.0      0.0   
3  0.088481  0.015789  ...       0.0       0.0      0.0      0.0      0.0   
4  0.088481  0.015789  ...       0.0       0.0      0.0      0.0      0.0   

   X5_X5-6  X5_X5-7  X5_X5-8  X5_X5-9            Y  
0      0.0      0.0      0.0      0.0  1546.836248  
1      0.0      0.0      0.0      0.0  1261.722593  
2      0.0      0.0      0.0      0.0   860.853275  
3      0.0      0.0      0.0      0.0   887.356349  
4      0.0      0.0      0.0      0.0   895.059756  

[5 rows x 50 columns]

In [13]:
import numpy as np
def split_train_test(data, train_ratio, random_seed):
    # to make this notebook's output identical at every run
    np.random.seed(random_seed)
    shuffled_indices = np.random.permutation(len(data))
    train_set_size = int(len(data) * train_ratio)
    # partition
    train_indices = shuffled_indices[:train_set_size]
    test_indices = shuffled_indices[train_set_size:]
    return data.iloc[train_indices], data.iloc[test_indices]

train_data, test_data = split_train_test(data, 0.75 , 128 )

train_data.head(2)

X6        X7        X8        X9       X10       X11       X12  \
1232  0.002395  0.073801  0.072243  0.091429  0.026316  0.000000  0.092593   
2271  0.016920  0.164207  0.174905  0.291429  0.105263  0.105263  0.277778   

           X13       X14       X15  ...  X5_X5-21  X5_X5-22  X5_X5-3  X5_X5-4  \
1232  0.000000  0.666110  0.015789  ...       0.0       0.0      0.0      0.0   
2271  0.073107  0.065109  0.100000  ...       0.0       0.0      0.0      0.0   

      X5_X5-5  X5_X5-6  X5_X5-7  X5_X5-8  X5_X5-9            Y  
1232      0.0      0.0      0.0      0.0      0.0  1307.817487  
2271      0.0      0.0      0.0      0.0      0.0   398.908415  

[2 rows x 50 columns]

In [14]:
corr_matrix = train_data.corr()
print(corr_matrix)

                X6        X7        X8        X9       X10       X11  \
X6        1.000000  0.018762 -0.134181 -0.042729  0.357182  0.166759   
X7        0.018762  1.000000  0.590991 -0.036867  0.518666  0.722287   
X8       -0.134181  0.590991  1.000000  0.248652  0.235019  0.585305   
X9       -0.042729 -0.036867  0.248652  1.000000  0.057629  0.199473   
X10       0.357182  0.518666  0.235019  0.057629  1.000000  0.769093   
X11       0.166759  0.722287  0.585305  0.199473  0.769093  1.000000   
X12       0.164090  0.528802  0.543090  0.351492  0.549801  0.765718   
X13       0.297680  0.563607  0.476013  0.030527  0.380147  0.672183   
X14      -0.234916 -0.537753 -0.515278 -0.174384 -0.686586 -0.768959   
X15       0.330360  0.491966  0.362986  0.058932  0.637129  0.694751   
X16       0.064682 -0.270442 -0.328241 -0.070832  0.071704 -0.122963   
X1_X1-1   0.000948 -0.514958 -0.622487 -0.169023 -0.256702 -0.674788   
X1_X1-2  -0.000948  0.514958  0.622487  0.169023  0.256702  0.67

In [15]:
corr_matrix[ "Y" ]

X6         -0.376066
X7         -0.520050
X8         -0.448213
X9         -0.179513
X10        -0.667452
X11        -0.747957
X12        -0.607864
X13        -0.640153
X14         0.709622
X15        -0.709226
X16         0.054002
X1_X1-1     0.483213
X1_X1-2    -0.483213
X2_X2-1    -0.029614
X2_X2-2     0.073223
X2_X2-3    -0.094132
X3_X3-1     0.098154
X3_X3-2     0.031472
X3_X3-3    -0.154302
X3_X3-4    -0.015720
X4_X4-1    -0.003149
X4_X4-2     0.225506
X4_X4-3    -0.041472
X4_X4-4     0.023204
X4_X4-5    -0.078723
X4_X4-6    -0.178767
X4_X4-7    -0.085931
X5_X5-1     0.710433
X5_X5-10   -0.017440
X5_X5-11    0.110300
X5_X5-12    0.169055
X5_X5-13   -0.052828
X5_X5-14   -0.016077
X5_X5-15   -0.062968
X5_X5-16   -0.197441
X5_X5-17   -0.248466
X5_X5-18    0.080516
X5_X5-19         NaN
X5_X5-2    -0.362699
X5_X5-20         NaN
X5_X5-21   -0.075854
X5_X5-22    0.117818
X5_X5-3    -0.033592
X5_X5-4    -0.007278
X5_X5-5    -0.025232
X5_X5-6    -0.019322
X5_X5-7    -0.010116
X5_X5-8    -0

In [16]:
import numpy as np
Y_corr = corr_matrix[ "Y" ][ (-np.abs(corr_matrix[ "Y" ])).argsort() ] 
Y_corr

X5_X5-8    -0.024401
X11        -0.747957
X5_X5-1     0.710433
X14         0.709622
X15        -0.709226
X10        -0.667452
X13        -0.640153
X12        -0.607864
X7         -0.520050
X1_X1-1     0.483213
X1_X1-2    -0.483213
X8         -0.448213
X6         -0.376066
X5_X5-19         NaN
X5_X5-17   -0.248466
X4_X4-2     0.225506
X5_X5-16   -0.197441
X9         -0.179513
X4_X4-6    -0.178767
X5_X5-12    0.169055
X3_X3-3    -0.154302
X5_X5-20         NaN
X5_X5-11    0.110300
X3_X3-1     0.098154
X2_X2-3    -0.094132
X4_X4-7    -0.085931
X5_X5-7    -0.010116
X5_X5-18    0.080516
X4_X4-5    -0.078723
X5_X5-2    -0.362699
X2_X2-2     0.073223
X5_X5-15   -0.062968
X16         0.054002
X5_X5-13   -0.052828
X4_X4-3    -0.041472
X5_X5-21   -0.075854
X3_X3-2     0.031472
Y           1.000000
X2_X2-1    -0.029614
Y           1.000000
X5_X5-3    -0.033592
X5_X5-6    -0.019322
X4_X4-4     0.023204
X5_X5-4    -0.007278
X5_X5-10   -0.017440
X5_X5-14   -0.016077
X3_X3-4    -0.015720
X5_X5-5    -0

* Q1: What's the most important correlation coefficients in terms of absolute magnitude with respect to the column Y.
    * Ans: X11      

* Q2: Does it make a difference as compared with the numerical features only? 
    * Ans: After adding categorical features, the most significant positive correlation shifts from X14 to X5_X5-1. However, in terms of negative correlation, X11 still maintains its maximum influence.

## (c)
* After dropping non-significant variables at once for the liner regression model, what are the training and testing metrics of RMSE and r2_score for the new linear regression model? 
* Are the performance indexes improved as compared with those of the numerical features only?

In [17]:
import statsmodels.api as sm
X_train = train_data.iloc[:, : 11]
Y_train = train_data.iloc[:, 11]
X_train_aug = sm.add_constant(X_train) # Add coefficients for the intercept to become augmented matrix
est = sm.OLS(Y_train, X_train_aug).fit() 
# Ordinary least squares (最小平方法)
print(est.summary())

                            OLS Regression Results                            
Dep. Variable:                X1_X1-1   R-squared:                       0.896
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     7711.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:00   Log-Likelihood:                 4614.2
No. Observations:                9888   AIC:                            -9204.
Df Residuals:                    9876   BIC:                            -9118.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9425      0.009    104.211      0.0

In [18]:
X_train_signif = pd.DataFrame(X_train, columns = throughput.columns[ 6 :]).drop("X7", axis= 1)
X_train_signif.head()

X6        X8        X9       X10       X11       X12       X13  \
1232   0.002395  0.072243  0.091429  0.026316  0.000000  0.092593  0.000000   
2271   0.016920  0.174905  0.291429  0.105263  0.105263  0.277778  0.073107   
1455   0.002559  0.077947  0.085714  0.026316  0.000000  0.092593  0.000000   
804    0.008010  0.319392  0.214286  0.157895  0.105263  0.216049  0.000000   
10163  0.015305  0.338403  0.255714  0.289474  0.342105  0.493827  0.409922   

            X14       X15       X16  
1232   0.666110  0.015789  0.166667  
2271   0.065109  0.100000  0.166667  
1455   0.666110  0.015789  0.166667  
804    0.088481  0.015789  0.333333  
10163  0.008347  0.263158  0.666667

In [19]:
X_train_signif = pd.DataFrame(X_train, columns = throughput.columns[ 6 :]).drop("X7", axis= 1)
X_train_signif = sm.add_constant(X_train_signif) # Add coefficients for the intercept
est_significant = sm.OLS(Y_train, X_train_signif.values).fit() 
# Ordinary least squares (最小平方法)
print( est_significant.summary() )

                            OLS Regression Results                            
Dep. Variable:                X1_X1-1   R-squared:                       0.894
Model:                            OLS   Adj. R-squared:                  0.894
Method:                 Least Squares   F-statistic:                     8321.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:00   Log-Likelihood:                 4528.9
No. Observations:                9888   AIC:                            -9036.
Df Residuals:                    9877   BIC:                            -8957.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9600      0.009    106.387      0.0

In [20]:
from statsmodels.tools.eval_measures import rmse 
from sklearn.metrics import r2_score
# https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html
print('train RMSE: ', round(rmse( Y_train , est_significant.predict(X_train_signif) ), 4)) 
print('train R^2: ', round(r2_score( Y_train, est_significant.predict(X_train_signif) ), 4)) 


# Need to drop nonsignificant columns from test_data
X3_test = pd.DataFrame(test_data.iloc[:, : ], columns = throughput.columns[ 6 :]).drop("X7", axis= 1)
X3_test = sm.add_constant(X3_test) # Add coefficients for the intercept
print('test RMSE: ',  round(rmse( test_data['Y'] , est_significant.predict(X3_test) ), 4)) 
print('test R^2: ', round(r2_score( test_data['Y'], est_significant.predict(X3_test) ), 4)) 

train RMSE:  0.1531
train R^2:  0.8939
test RMSE:  672.3261
test R^2:  -1.1393


In [21]:
est.rsquared

0.8957135068680173

In [22]:
est_significant.rsquared

0.8938993723427809

* Q1: After dropping non-significant variables, what are the training and testing metrics of RMSE and r2_score for the new linear regression model? 
    * Ans:
        * train RMSE: 245.9784
        * train R^2: 0.7161
        * test RMSE: 252.0665
        * test R^2: 0.6993

* Q2: Are the performance indexes improved as compared with those of the numerical features only?
    * Ans: Yes, it improved from 0.7161082629542501 to 0.7160983210526781. 

## (d)
* If you add nonlinear features, e.g., 𝑥2, √𝑥, or log( 𝑥 + 0.00001) for each feature, do your RMSE and R2 scores improve over those of (c)? 
* If you use 3 nonlinear feature for each feature, then you will have 4 features for each one. In this case, data exploration might help to gain insight about the format of your nonlinear functions. You could write a function to automate the process of generating many features from 1 feature or the whole dataframe. 
    * (Hint: For a given feature vector x, the command to generate a nonlinear features is to use NumPy function, e.g. np.sqrt(x). For a given dataframe df, you need to use df.columns to access its column names.)

In [23]:
throughput.head()

Y    X1    X2    X3    X4    X5        X6   X7   X8   X9  X10  \
0  1546.836248  X1-1  X2-1  X3-1  X4-1  X5-1  0.025286   50   48  105  3.0   
1  1261.722593  X1-1  X2-1  X3-1  X4-1  X5-1  0.034876   67   34  105  3.0   
2   860.853275  X1-1  X2-1  X3-1  X4-1  X5-1  0.079237  176  187  200  8.0   
3   887.356349  X1-1  X2-1  X3-1  X4-1  X5-1  0.113288  176  187  200  8.0   
4   895.059756  X1-1  X2-1  X3-1  X4-1  X5-1  0.113778  176  187  200  8.0   

   X11   X12  X13   X14  X15  X16  
0  2.0  0.45    8  1600    8  0.8  
1  2.0  0.45    8  1600   10  0.8  
2  6.0  0.75    8   216   10  0.8  
3  6.0  0.75    8   216   10  0.8  
4  6.0  0.75    8   216   10  0.8

In [24]:
throughput.iloc[:, 6:]

X6   X7   X8   X9   X10   X11   X12  X13   X14  X15  X16
0      0.025286   50   48  105   3.0   2.0  0.45    8  1600    8  0.8
1      0.034876   67   34  105   3.0   2.0  0.45    8  1600   10  0.8
2      0.079237  176  187  200   8.0   6.0  0.75    8   216   10  0.8
3      0.113288  176  187  200   8.0   6.0  0.75    8   216   10  0.8
4      0.113778  176  187  200   8.0   6.0  0.75    8   216   10  0.8
...         ...  ...  ...  ...   ...   ...   ...  ...   ...  ...  ...
13181  0.149699  479  247   60  12.0  18.0  1.20  132    60   88  0.8
13182  0.300309  217  216   80   8.0  10.5  1.00  162   168  126  0.7
13183  0.300309  217  216   80   8.0  10.5  1.00  162   168   26  0.7
13184  0.154297  370  327  150  10.0  14.5  0.80  200    70  182  0.7
13185  0.163640  370  327  150  10.0  14.5  0.80  200    70  182  0.7

[13185 rows x 11 columns]

In [25]:
throughput_new = throughput.iloc[:, 6:].copy()
throughput_new

X6   X7   X8   X9   X10   X11   X12  X13   X14  X15  X16
0      0.025286   50   48  105   3.0   2.0  0.45    8  1600    8  0.8
1      0.034876   67   34  105   3.0   2.0  0.45    8  1600   10  0.8
2      0.079237  176  187  200   8.0   6.0  0.75    8   216   10  0.8
3      0.113288  176  187  200   8.0   6.0  0.75    8   216   10  0.8
4      0.113778  176  187  200   8.0   6.0  0.75    8   216   10  0.8
...         ...  ...  ...  ...   ...   ...   ...  ...   ...  ...  ...
13181  0.149699  479  247   60  12.0  18.0  1.20  132    60   88  0.8
13182  0.300309  217  216   80   8.0  10.5  1.00  162   168  126  0.7
13183  0.300309  217  216   80   8.0  10.5  1.00  162   168   26  0.7
13184  0.154297  370  327  150  10.0  14.5  0.80  200    70  182  0.7
13185  0.163640  370  327  150  10.0  14.5  0.80  200    70  182  0.7

[13185 rows x 11 columns]

In [26]:
for i in range(6, 17):
    throughput_new["X" + str(i) + "_2"] = throughput_new["X" + str(i)] ** 2

throughput_new

X6   X7   X8   X9   X10   X11   X12  X13   X14  X15  ...    X7_2  \
0      0.025286   50   48  105   3.0   2.0  0.45    8  1600    8  ...    2500   
1      0.034876   67   34  105   3.0   2.0  0.45    8  1600   10  ...    4489   
2      0.079237  176  187  200   8.0   6.0  0.75    8   216   10  ...   30976   
3      0.113288  176  187  200   8.0   6.0  0.75    8   216   10  ...   30976   
4      0.113778  176  187  200   8.0   6.0  0.75    8   216   10  ...   30976   
...         ...  ...  ...  ...   ...   ...   ...  ...   ...  ...  ...     ...   
13181  0.149699  479  247   60  12.0  18.0  1.20  132    60   88  ...  229441   
13182  0.300309  217  216   80   8.0  10.5  1.00  162   168  126  ...   47089   
13183  0.300309  217  216   80   8.0  10.5  1.00  162   168   26  ...   47089   
13184  0.154297  370  327  150  10.0  14.5  0.80  200    70  182  ...  136900   
13185  0.163640  370  327  150  10.0  14.5  0.80  200    70  182  ...  136900   

         X8_2   X9_2  X10_2   X11_2   X12_2  X13_2    X14_2  X15_2  X16_2  
0        2304  11025    9.0    4.00  0.2025     64  2560000     64   0.64  
1        1156  11025    9.0    4.00  0.2025     64  2560000    100   0.64  
2       34969  40000   64.0   36.00  0.5625     64    46656    100   0.64  
3       34969  40000   64.0   36.00  0.5625     64    46656    100   0.64  
4       34969  40000   64.0   36.00  0.5625     64    46656    100   0.64  
...       ...    ...    ...     ...     ...    ...      ...    ...    ...  
13181   61009   3600  144.0  324.00  1.4400  17424     3600   7744   0.64  
13182   46656   6400   64.0  110.25  1.0000  26244    28224  15876   0.49  
13183   46656   6400   64.0  110.25  1.0000  26244    28224    676   0.49  
13184  106929  22500  100.0  210.25  0.6400  40000     4900  33124   0.49  
13185  106929  22500  100.0  210.25  0.6400  40000     4900  33124   0.49  

[13185 rows x 22 columns]

In [27]:
from sklearn.preprocessing import MinMaxScaler
ss = MinMaxScaler()
X2_scaled = ss.fit_transform(throughput_new)
X2_scaled.shape

(13185, 22)

In [28]:
new_data = pd.concat([pd.DataFrame(X2_scaled, columns = throughput_new.columns[ 0 :]), 
                  pd.DataFrame(throughput['Y'].values, columns = [throughput.columns[0]])], axis = 1) 
new_data.head(2) # See the result and make sure that the above command works!

X6        X7        X8        X9       X10  X11       X12  X13  \
0  0.001773  0.057196  0.055133  0.078571  0.026316  0.0  0.030864  0.0   
1  0.002447  0.088561  0.028517  0.078571  0.026316  0.0  0.030864  0.0   

       X14       X15  ...      X8_2      X9_2     X10_2  X11_2     X12_2  \
0  0.66611  0.010526  ...  0.006549  0.015223  0.003133    0.0  0.010841   
1  0.66611  0.015789  ...  0.002680  0.015223  0.003133    0.0  0.010841   

   X13_2     X14_2     X15_2     X16_2            Y  
0    0.0  0.444443  0.000326  0.259259  1546.836248  
1    0.0  0.444443  0.000570  0.259259  1261.722593  

[2 rows x 23 columns]

In [29]:
import numpy as np
def split_train_test(data, train_ratio, random_seed):
    # to make this notebook's output identical at every run
    np.random.seed(random_seed)
    shuffled_indices = np.random.permutation(len(data))
    train_set_size = int(len(data) * train_ratio)
    # partition
    train_indices = shuffled_indices[:train_set_size]
    test_indices = shuffled_indices[train_set_size:]
    return data.iloc[train_indices], data.iloc[test_indices]

train2_data, test2_data = split_train_test(new_data, 0.75 , 128 )

train2_data.head(2)

X6        X7        X8        X9       X10       X11       X12  \
1232  0.002395  0.073801  0.072243  0.091429  0.026316  0.000000  0.092593   
2271  0.016920  0.164207  0.174905  0.291429  0.105263  0.105263  0.277778   

           X13       X14       X15  ...      X8_2      X9_2     X10_2  \
1232  0.000000  0.666110  0.015789  ...  0.009735  0.018743  0.003133   
2271  0.073107  0.065109  0.100000  ...  0.040315  0.110743  0.020050   

        X11_2     X12_2     X13_2     X14_2     X15_2    X16_2            Y  
1232  0.00000  0.036348  0.000000  0.444443  0.000570  0.12037  1307.817487  
2271  0.02005  0.143480  0.008062  0.004442  0.011856  0.12037   398.908415  

[2 rows x 23 columns]

In [30]:
import statsmodels.api as sm
X_train = train2_data.iloc[:, : 22]
Y_train = train2_data.iloc[:, 22]
X_train_aug = sm.add_constant(X_train) # Add coefficients for the intercept to become augmented matrix
est = sm.OLS(Y_train, X_train_aug).fit() 
# Ordinary least squares (最小平方法)
print(est.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.800
Method:                 Least Squares   F-statistic:                     1794.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:01   Log-Likelihood:                -66733.
No. Observations:                9888   AIC:                         1.335e+05
Df Residuals:                    9865   BIC:                         1.337e+05
Df Model:                          22                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        887.0725     39.452     22.485      0.0

In [31]:
p_value_thres = 0.05 # threshold
nonsignificant = [] # A list of nonsignificant variables. It is empty initially.
significant = throughput_new.columns[ 0 :].tolist() # The initial significant variables
for i in range(len(throughput_new.columns[ 0 :])): 
    if est.pvalues[i+1] >= p_value_thres:
        # Add the nonsignificant variable to the list of nonsignificant variables
        nonsignificant.append(throughput_new.columns[ 0 +i]) 
        # Remove the nonsignificant variable out of the list of significant variables
        significant.remove(throughput_new.columns[ 0 +i])

print(nonsignificant, significant)

['X8', 'X11', 'X7_2', 'X8_2', 'X10_2'] ['X6', 'X7', 'X9', 'X10', 'X12', 'X13', 'X14', 'X15', 'X16', 'X6_2', 'X9_2', 'X11_2', 'X12_2', 'X13_2', 'X14_2', 'X15_2', 'X16_2']


In [32]:
X_train_signif = pd.DataFrame(X_train, columns = throughput_new.columns[ 0 :]).drop(nonsignificant, axis = 1)
X_train_signif.head(2)

X6        X7        X9       X10       X12       X13       X14  \
1232  0.002395  0.073801  0.091429  0.026316  0.092593  0.000000  0.666110   
2271  0.016920  0.164207  0.291429  0.105263  0.277778  0.073107  0.065109   

           X15       X16      X6_2      X9_2    X11_2     X12_2     X13_2  \
1232  0.015789  0.166667  0.000006  0.018743  0.00000  0.036348  0.000000   
2271  0.100000  0.166667  0.000286  0.110743  0.02005  0.143480  0.008062   

         X14_2     X15_2    X16_2  
1232  0.444443  0.000570  0.12037  
2271  0.004442  0.011856  0.12037

In [33]:
from sklearn.linear_model import LinearRegression
  
lr = LinearRegression()
lr.fit(X_train_signif,Y_train)
# LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
#     normalize=False)

lr.score(X_train_signif,Y_train)

0.7998601023377987

In [34]:
X_train_signif = pd.DataFrame(X_train, columns = throughput_new.columns[ 0 :]).drop(nonsignificant, axis= 1)
X_train_signif = sm.add_constant(X_train_signif) # Add coefficients for the intercept
est_significant = sm.OLS(Y_train, X_train_signif.values).fit() 
# Ordinary least squares (最小平方法)
print( est_significant.summary() )

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.800
Method:                 Least Squares   F-statistic:                     2320.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:01   Log-Likelihood:                -66738.
No. Observations:                9888   AIC:                         1.335e+05
Df Residuals:                    9870   BIC:                         1.336e+05
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        888.2869     30.605     29.024      0.0

In [35]:
from statsmodels.tools.eval_measures import rmse 
from sklearn.metrics import r2_score
# https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html
print('train RMSE: ', round(rmse( Y_train, est_significant.predict( X_train_signif ) ), 4)) 
print('train R^2: ', round(r2_score( Y_train, est_significant.predict( X_train_signif ) ), 4)) 


# Need to drop nonsignificant columns from test_data
X3_test = pd.DataFrame(test2_data.iloc[:, : ], columns = throughput_new.columns[ 0 :]).drop(nonsignificant, axis= 1)
X3_test = sm.add_constant(X3_test) # Add coefficients for the intercept
print('test RMSE: ',  round(rmse( test2_data['Y'], est_significant.predict(X3_test) ), 4)) 
print('test R^2: ', round(r2_score( test2_data['Y'], est_significant.predict(X3_test )), 4)) 

train RMSE:  206.5284
train R^2:  0.7999
test RMSE:  218.1844
test R^2:  0.7747


* Q1: If you add nonlinear features, e.g., 𝑥2, √𝑥, or log( 𝑥 + 0.00001) for each feature, do your RMSE and R2 scores improve over those of (c)?
    * Ans: 
        * The former result:<br>
            train RMSE:  245.9784<br>
            train R^2:  0.7161<br>
            test RMSE:  252.0665<br>
            test R^2:  0.6993<br>
        * Change to:<br>
            train RMSE:  206.5284<br>
            train R^2:  0.7999<br>
            test RMSE:  218.1844<br>
            test R^2:  0.7747<br>

In [36]:
for i in range(6, 17):
    throughput_new["X" + str(i) + "_3"] = throughput_new["X" + str(i)] ** 3

throughput_new

X6   X7   X8   X9   X10   X11   X12  X13   X14  X15  ...  \
0      0.025286   50   48  105   3.0   2.0  0.45    8  1600    8  ...   
1      0.034876   67   34  105   3.0   2.0  0.45    8  1600   10  ...   
2      0.079237  176  187  200   8.0   6.0  0.75    8   216   10  ...   
3      0.113288  176  187  200   8.0   6.0  0.75    8   216   10  ...   
4      0.113778  176  187  200   8.0   6.0  0.75    8   216   10  ...   
...         ...  ...  ...  ...   ...   ...   ...  ...   ...  ...  ...   
13181  0.149699  479  247   60  12.0  18.0  1.20  132    60   88  ...   
13182  0.300309  217  216   80   8.0  10.5  1.00  162   168  126  ...   
13183  0.300309  217  216   80   8.0  10.5  1.00  162   168   26  ...   
13184  0.154297  370  327  150  10.0  14.5  0.80  200    70  182  ...   
13185  0.163640  370  327  150  10.0  14.5  0.80  200    70  182  ...   

            X7_3      X8_3     X9_3   X10_3     X11_3     X12_3    X13_3  \
0         125000    110592  1157625    27.0     8.000  0.091125      512   
1         300763     39304  1157625    27.0     8.000  0.091125      512   
2        5451776   6539203  8000000   512.0   216.000  0.421875      512   
3        5451776   6539203  8000000   512.0   216.000  0.421875      512   
4        5451776   6539203  8000000   512.0   216.000  0.421875      512   
...          ...       ...      ...     ...       ...       ...      ...   
13181  109902239  15069223   216000  1728.0  5832.000  1.728000  2299968   
13182   10218313  10077696   512000   512.0  1157.625  1.000000  4251528   
13183   10218313  10077696   512000   512.0  1157.625  1.000000  4251528   
13184   50653000  34965783  3375000  1000.0  3048.625  0.512000  8000000   
13185   50653000  34965783  3375000  1000.0  3048.625  0.512000  8000000   

            X14_3    X15_3  X16_3  
0      4096000000      512  0.512  
1      4096000000     1000  0.512  
2        10077696     1000  0.512  
3        10077696     1000  0.512  
4        10077696     1000  0.512  
...           ...      ...    ...  
13181      216000   681472  0.512  
13182     4741632  2000376  0.343  
13183     4741632    17576  0.343  
13184      343000  6028568  0.343  
13185      343000  6028568  0.343  

[13185 rows x 33 columns]

In [37]:
for i in range(6, 17):
    throughput_new["X" + str(i) + "_sqrt"] = throughput_new["X" + str(i)] * np.sqrt(2)


throughput_new

X6   X7   X8   X9   X10   X11   X12  X13   X14  X15  ...  \
0      0.025286   50   48  105   3.0   2.0  0.45    8  1600    8  ...   
1      0.034876   67   34  105   3.0   2.0  0.45    8  1600   10  ...   
2      0.079237  176  187  200   8.0   6.0  0.75    8   216   10  ...   
3      0.113288  176  187  200   8.0   6.0  0.75    8   216   10  ...   
4      0.113778  176  187  200   8.0   6.0  0.75    8   216   10  ...   
...         ...  ...  ...  ...   ...   ...   ...  ...   ...  ...  ...   
13181  0.149699  479  247   60  12.0  18.0  1.20  132    60   88  ...   
13182  0.300309  217  216   80   8.0  10.5  1.00  162   168  126  ...   
13183  0.300309  217  216   80   8.0  10.5  1.00  162   168   26  ...   
13184  0.154297  370  327  150  10.0  14.5  0.80  200    70  182  ...   
13185  0.163640  370  327  150  10.0  14.5  0.80  200    70  182  ...   

          X7_sqrt     X8_sqrt     X9_sqrt   X10_sqrt   X11_sqrt  X12_sqrt  \
0       70.710678   67.882251  148.492424   4.242641   2.828427  0.636396   
1       94.752309   48.083261  148.492424   4.242641   2.828427  0.636396   
2      248.901587  264.457936  282.842712  11.313708   8.485281  1.060660   
3      248.901587  264.457936  282.842712  11.313708   8.485281  1.060660   
4      248.901587  264.457936  282.842712  11.313708   8.485281  1.060660   
...           ...         ...         ...        ...        ...       ...   
13181  677.408296  349.310750   84.852814  16.970563  25.455844  1.697056   
13182  306.884343  305.470129  113.137085  11.313708  14.849242  1.414214   
13183  306.884343  305.470129  113.137085  11.313708  14.849242  1.414214   
13184  523.259018  462.447835  212.132034  14.142136  20.506097  1.131371   
13185  523.259018  462.447835  212.132034  14.142136  20.506097  1.131371   

         X13_sqrt     X14_sqrt    X15_sqrt  X16_sqrt  
0       11.313708  2262.741700   11.313708  1.131371  
1       11.313708  2262.741700   14.142136  1.131371  
2       11.313708   305.470129   14.142136  1.131371  
3       11.313708   305.470129   14.142136  1.131371  
4       11.313708   305.470129   14.142136  1.131371  
...           ...          ...         ...       ...  
13181  186.676190    84.852814  124.450793  1.131371  
13182  229.102597   237.587878  178.190909  0.989949  
13183  229.102597   237.587878   36.769553  0.989949  
13184  282.842712    98.994949  257.386868  0.989949  
13185  282.842712    98.994949  257.386868  0.989949  

[13185 rows x 44 columns]

In [38]:
from sklearn.preprocessing import MinMaxScaler
ss = MinMaxScaler()
X2_scaled = ss.fit_transform(throughput_new)
X2_scaled.shape

(13185, 44)

In [39]:
new_data = pd.concat([pd.DataFrame(X2_scaled, columns = throughput_new.columns[ 0 :]), 
                  pd.DataFrame(throughput['Y'].values, columns = [throughput.columns[0]])], axis = 1) 
new_data.head(2) # See the result and make sure that the above command works!

X6        X7        X8        X9       X10  X11       X12  X13  \
0  0.001773  0.057196  0.055133  0.078571  0.026316  0.0  0.030864  0.0   
1  0.002447  0.088561  0.028517  0.078571  0.026316  0.0  0.030864  0.0   

       X14       X15  ...   X8_sqrt   X9_sqrt  X10_sqrt  X11_sqrt  X12_sqrt  \
0  0.66611  0.010526  ...  0.055133  0.078571  0.026316       0.0  0.030864   
1  0.66611  0.015789  ...  0.028517  0.078571  0.026316       0.0  0.030864   

   X13_sqrt  X14_sqrt  X15_sqrt  X16_sqrt            Y  
0       0.0   0.66611  0.010526  0.333333  1546.836248  
1       0.0   0.66611  0.015789  0.333333  1261.722593  

[2 rows x 45 columns]

In [40]:
import numpy as np
def split_train_test(data, train_ratio, random_seed):
    # to make this notebook's output identical at every run
    np.random.seed(random_seed)
    shuffled_indices = np.random.permutation(len(data))
    train_set_size = int(len(data) * train_ratio)
    # partition
    train_indices = shuffled_indices[:train_set_size]
    test_indices = shuffled_indices[train_set_size:]
    return data.iloc[train_indices], data.iloc[test_indices]

train2_data, test2_data = split_train_test(new_data, 0.75 , 128 )

train2_data.head(2)

X6        X7        X8        X9       X10       X11       X12  \
1232  0.002395  0.073801  0.072243  0.091429  0.026316  0.000000  0.092593   
2271  0.016920  0.164207  0.174905  0.291429  0.105263  0.105263  0.277778   

           X13       X14       X15  ...   X8_sqrt   X9_sqrt  X10_sqrt  \
1232  0.000000  0.666110  0.015789  ...  0.072243  0.091429  0.026316   
2271  0.073107  0.065109  0.100000  ...  0.174905  0.291429  0.105263   

      X11_sqrt  X12_sqrt  X13_sqrt  X14_sqrt  X15_sqrt  X16_sqrt            Y  
1232  0.000000  0.092593  0.000000  0.666110  0.015789  0.166667  1307.817487  
2271  0.105263  0.277778  0.073107  0.065109  0.100000  0.166667   398.908415  

[2 rows x 45 columns]

In [41]:
import statsmodels.api as sm
X_train = train2_data.iloc[:, : -1]
Y_train = train2_data.iloc[:, -1]
X_train_aug = sm.add_constant(X_train) # Add coefficients for the intercept to become augmented matrix
est = sm.OLS(Y_train, X_train_aug).fit() 
# Ordinary least squares (最小平方法)
print(est.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.843
Model:                            OLS   Adj. R-squared:                  0.842
Method:                 Least Squares   F-statistic:                     1598.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:01   Log-Likelihood:                -65552.
No. Observations:                9888   AIC:                         1.312e+05
Df Residuals:                    9854   BIC:                         1.314e+05
Df Model:                          33                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        890.7742     63.582     14.010      0.0

In [42]:
p_value_thres = 0.05 # threshold
nonsignificant = [] # A list of nonsignificant variables. It is empty initially.
significant = throughput_new.columns[ 0 :].tolist() # The initial significant variables
for i in range(len(throughput_new.columns[ 0 :])): 
    if est.pvalues[i+1] >= p_value_thres:
        # Add the nonsignificant variable to the list of nonsignificant variables
        nonsignificant.append(throughput_new.columns[ 0 +i]) 
        # Remove the nonsignificant variable out of the list of significant variables
        significant.remove(throughput_new.columns[ 0 +i])

print(nonsignificant, significant)

['X8', 'X9', 'X8_2', 'X9_2', 'X7_3', 'X8_3', 'X8_sqrt', 'X9_sqrt'] ['X6', 'X7', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X6_2', 'X7_2', 'X10_2', 'X11_2', 'X12_2', 'X13_2', 'X14_2', 'X15_2', 'X16_2', 'X6_3', 'X9_3', 'X10_3', 'X11_3', 'X12_3', 'X13_3', 'X14_3', 'X15_3', 'X16_3', 'X6_sqrt', 'X7_sqrt', 'X10_sqrt', 'X11_sqrt', 'X12_sqrt', 'X13_sqrt', 'X14_sqrt', 'X15_sqrt', 'X16_sqrt']


In [43]:
X_train_signif = pd.DataFrame(X_train, columns = throughput_new.columns[ 0 :]).drop(nonsignificant, axis = 1)
X_train_signif.head(2)

X6        X7       X10       X11       X12       X13       X14  \
1232  0.002395  0.073801  0.026316  0.000000  0.092593  0.000000  0.666110   
2271  0.016920  0.164207  0.105263  0.105263  0.277778  0.073107  0.065109   

           X15       X16      X6_2  ...     X16_3   X6_sqrt   X7_sqrt  \
1232  0.015789  0.166667  0.000006  ...  0.083995  0.002395  0.073801   
2271  0.100000  0.166667  0.000286  ...  0.083995  0.016920  0.164207   

      X10_sqrt  X11_sqrt  X12_sqrt  X13_sqrt  X14_sqrt  X15_sqrt  X16_sqrt  
1232  0.026316  0.000000  0.092593  0.000000  0.666110  0.015789  0.166667  
2271  0.105263  0.105263  0.277778  0.073107  0.065109  0.100000  0.166667  

[2 rows x 36 columns]

In [44]:
from sklearn.linear_model import LinearRegression
  
lr = LinearRegression()
lr.fit(X_train_signif,Y_train)
# LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
#     normalize=False)

lr.score(X_train_signif,Y_train)

0.838668439491531

In [45]:
X_train_signif = pd.DataFrame(X_train, columns = throughput_new.columns[ 0 :]).drop(nonsignificant, axis= 1)
X_train_signif = sm.add_constant(X_train_signif) # Add coefficients for the intercept
est_significant = sm.OLS(Y_train, X_train_signif.values).fit() 
# Ordinary least squares (最小平方法)
print( est_significant.summary() )

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.839
Model:                            OLS   Adj. R-squared:                  0.839
Method:                 Least Squares   F-statistic:                     1904.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:02   Log-Likelihood:                -65659.
No. Observations:                9888   AIC:                         1.314e+05
Df Residuals:                    9860   BIC:                         1.316e+05
Df Model:                          27                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        954.6126     60.566     15.762      0.0

In [46]:
from statsmodels.tools.eval_measures import rmse 
from sklearn.metrics import r2_score
# https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html
print('train RMSE: ', round(rmse( Y_train, est_significant.predict( X_train_signif ) ), 4)) 
print('train R^2: ', round(r2_score( Y_train, est_significant.predict( X_train_signif ) ), 4)) 


# Need to drop nonsignificant columns from test_data
X3_test = pd.DataFrame(test2_data.iloc[:, : ], columns = throughput_new.columns[ 0 :]).drop(nonsignificant, axis= 1)
X3_test = sm.add_constant(X3_test) # Add coefficients for the intercept
print('test RMSE: ',  round(rmse( test2_data['Y'], est_significant.predict(X3_test) ), 4)) 
print('test R^2: ', round(r2_score( test2_data['Y'], est_significant.predict(X3_test )), 4)) 

train RMSE:  185.1833
train R^2:  0.8391
test RMSE:  196.2489
test R^2:  0.8177


* Q2: If you use 3 nonlinear feature for each feature, then you will have 4 features for each one. In this case, data exploration might help to gain insight about the format of your nonlinear functions. You could write a function to automate the process of generating many features from 1 feature or the whole dataframe. 
    * Ans: 
        * The former result:<br>
            train RMSE:  245.9784<br>
            train R^2:  0.7161<br>
            test RMSE:  252.0665<br>
            test R^2:  0.6993<br>
        * Add one nonlinear feature will change to:<br>
            train RMSE:  206.5284<br>
            train R^2:  0.7999<br>
            test RMSE:  218.1844<br>
            test R^2:  0.7747<br>
        * And add three nonlinear features will change to:<br>
            train RMSE:  185.1833<br>
            train R^2:  0.8391<br>
            test RMSE:  196.2489<br>
            test R^2:  0.8177<br>

## (e) Bonus
* For problem (c), please use the formal backward elimination procedure to eliminate non-significant variables.<br>
* In this case, you need to write a function which choose the most non-significant variable with the largest number and use for loop to call the function until there is no non-significant variable.

In [47]:
import statsmodels.api as sm
X_train = train_data.iloc[:, : 11]
Y_train = train_data.iloc[:, 11]
X_train_aug = sm.add_constant(X_train) # Add coefficients for the intercept to become augmented matrix
est = sm.OLS(Y_train, X_train_aug).fit() 
# Ordinary least squares (最小平方法)
print(est.summary())

                            OLS Regression Results                            
Dep. Variable:                X1_X1-1   R-squared:                       0.896
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     7711.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:02   Log-Likelihood:                 4614.2
No. Observations:                9888   AIC:                            -9204.
Df Residuals:                    9876   BIC:                            -9118.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9425      0.009    104.211      0.0

In [48]:
p_value_thres = 0.05 # threshold
nonsignificant = [] # A list of nonsignificant variables. It is empty initially.
significant = throughput.columns[ 6 :].tolist() # The initial significant variables
for i in range(len(throughput.columns[ 6 :])): 
    if est.pvalues[i+1] >= p_value_thres:
        # Add the nonsignificant variable to the list of nonsignificant variables
        nonsignificant.append(throughput.columns[ 6 +i]) 
        # Remove the nonsignificant variable out of the list of significant variables
        significant.remove(throughput.columns[ 6 +i])

print(nonsignificant, significant)

[] ['X6', 'X7', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16']


In [49]:
X_train_signif = pd.DataFrame(X_train, columns = throughput.columns[ 6 :]).drop(nonsignificant, axis= 1)
X_train_signif = sm.add_constant(X_train_signif) # Add coefficients for the intercept
est_significant = sm.OLS(Y_train, X_train_signif.values).fit() 
# Ordinary least squares (最小平方法)
print( est_significant.summary() )

                            OLS Regression Results                            
Dep. Variable:                X1_X1-1   R-squared:                       0.896
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     7711.
Date:                Wed, 18 Oct 2023   Prob (F-statistic):               0.00
Time:                        22:34:02   Log-Likelihood:                 4614.2
No. Observations:                9888   AIC:                            -9204.
Df Residuals:                    9876   BIC:                            -9118.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9425      0.009    104.211      0.0